File: python/015/hard_worker.py
from typing import List

# See https://colab.research.google.com/drive/1jjLSye6v7nXfwhPYx4t1IfXy1_lQvcz0
# N = jobs.length
# time_complexity O(N * 3 * 3 + 3) = O(N)
# space_complexity O(3:prev_totals + 3:current_totals) = O(1)


def solve(jobs: List[List[int]]) -> int:
    prev_totals = [0, 0, 0]

    for job_row in jobs:
        current_totals = [0, 0, 0]

        for job_id, reward in enumerate(job_row):
            current_totals[job_id] = _get_max_reward(prev_totals, job_id) + reward

        prev_totals = current_totals

    return max(prev_totals)


def _get_max_reward(job_row: List[int], ignore_id: int) -> int:
    max_reward = -1
    for job_id, reward in enumerate(job_row):
        if job_id == ignore_id:
            continue
        max_reward = max(max_reward, reward)

    if max_reward == -1:
        raise ValueError("Reward情報が取得できませんでした")

    return max_reward

File: src/plus/015/hard-worker.ts
/**
 * @see https://colab.research.google.com/drive/1jjLSye6v7nXfwhPYx4t1IfXy1_lQvcz0
 *
 * N = jobs.length
 * M = jobs[x].length = 3(Fixed value)
 * @time_complexity O(NM) = O(N)
 * @space_complexity O(3:prevTotals + 3:currentTotals) = O(1)
 */
export const solve = (jobs: number[][]): number => {
    let prevTotals: [number, number, number] = [0, 0, 0];
    for (let y = 0; y < jobs.length; y++) {
        const currentTotals: [number, number, number] = [0, 0, 0];
        const jobRow = jobs[y];

        for (let x = 0; x < jobRow.length; x++) {
            currentTotals[x] = getMaxReward(prevTotals, x) + jobRow[x];
        }

        prevTotals = currentTotals;
    }

    return Math.max(...prevTotals);
};

function getMaxReward(jobsRow: number[], ignoreId: number) {
    let maxRewards = -1;

    for (let x = 0; x < jobsRow.length; x++) {
        const reward = jobsRow[x];
        if (ignoreId === x) {
            continue;
        }
        if (maxRewards < reward) {
            maxRewards = reward;
        }
    }
    if (maxRewards === -1) {
        throw new Error('Reward情報が取得できませんでした');
    }
    return maxRewards;
}
